File: Leetcode/Array/015_3Sum.java
public class Solution {
    // O(n^2) time, O(1) space
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> ret = new LinkedList<List<Integer>>();
        if(nums == null || nums.length == 0)
            return ret;
        Arrays.sort(nums);
        
        for(int i = 0; i < nums.length - 2; i++){
            if(i == 0 || nums[i] != nums[i-1]){
                int l = i + 1, r = nums.length - 1;
                while(l < r){
                    if(nums[i] + nums[l] + nums[r] == 0){
                        List<Integer> sol = new LinkedList<Integer>();
                        sol.add(nums[i]);
                        sol.add(nums[l]);
                        sol.add(nums[r]);
                        ret.add(sol);
                        l++; r--;
                        while(l < r && nums[l] == nums[l-1])
                            l++;
                        while(l < r && nums[r] == nums[r+1])
                            r--;
                    }else if(nums[i] + nums[l] + nums[r] > 0){
                        r--;                    
                    }else{
                        l++;
                    }
                }    
            }
        }
        return ret;
    }
}

File: Leetcode/All/015_3Sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        vector<vector<int>> ret;
        if(nums.size() < 3) 
            return ret;
        sort(nums.begin(), nums.end());
        
        for(int i = 0; i < nums.size() - 2; i++){
            if(i != 0 && nums[i] == nums[i-1])  continue;
            int l = i + 1, r = nums.size() - 1;
            while(l < r){
                int sum = nums[i] + nums[r] + nums[l];
                if(sum == 0){
                    vector<int> sol;
                    sol.push_back(nums[i]);
                    sol.push_back(nums[l]);
                    sol.push_back(nums[r]);
                    ret.push_back(sol);
                    l++; r--;
                    while(l < r && nums[l] == nums[l-1])
                        l++;
                    while(l < r && nums[r] == nums[r+1])
                        r--;
                }else if(sum < 0){
                    l++;
                }else{
                    r--;
                }
            }
        }
        return ret;
    }
};